### ISSUE TITLE
frequent disconnects

### ISSUE THREAD

**Issue by brianmay** [dc297/mqttclpro#19]
Hello,

As much as a like this, it seems to disconnect frequently for no apparent reason. Still trying to work out why. I have tried changing the keep alive from 1200 seconds to 60, but it doesn't appear to have helped.

So far if I try to reproduce the problem on demand, it works perfectly :-(

When it does disconnect, it doesn't appear to retry to connect unless the screen is on. Although that could also be coincidence too.

This is on Android 7.1.2 connecting to mosquitto 1.3.4.

Regards

**Comment by brianmay** [comment]
"Waiting for network connection." seems to be common when the connection goes down (without reason). The retry time is set to 100 seconds, and sometimes there doesn't appear to be any attempt to reconnect after this time elapses.

**Comment by acefed** [comment]
I encountered the same issue, and am still trying to pinpoint the problem as well. My context:

- Using Android 7.0, Tasker v5.0u4, Mqttclpro v2.2.2 + 2 Samsung Galaxy S6 phones clientside;
- Broker is Mosquitto v1.4.14 on Raspberry Pi, all settings default;
- Mqttclpro keepalive was lowered to 300, Retry interval is 100, with user/pass only (no SSL). No clean session;
- Android power management is disabled for Tasker and Mqttclpro, including background services;
- I frequently change network (from Wifi to VPN), but it happens on the same local network as well. VPN is IPv4 only, all other connections IPv6;
- Subscriptions and publishing are impacted, the latter shows "Unable to publish as we are not connected";
- A different Tasker plugin [MQTT Publish Plugin](https://github.com/nosybore/Tasker-MQTT-Publish-Plugin) does publishing only, but does not seem to have this issue (perhaps their code/approach helps as an example?);
- Waking up Android shows "Connecting..." first and then "Connected", this often restores the connection correctly (more testing needed);

Planning to dive into the serverside Mosquitto settings next, but I have a feeling the source of the issue lies within Mqttclpro.

**Comment by brianmay** [comment]
Might be related to #14. Similar to what is discussed in this bug, it often doesn't seem to retry reconnecting when it says it will - or at least the display keeps saying it will retry at time X after time X has elapsed. When it reconnects correctly, I might be able to forgive it for disconnecting in the first place :-).

Possibly sometime disconnects are due to poor wifi reception when I am moving around, however, I have also had disconnects when the phone is sitting pretty much right next to the (very lowly loaded) wifi router, and there is no evidence that the wifi connection is being dropped (probably should double check this).

I am inclined to feel that I need run tcpdump on the connection, and see if that gives away any secrets. Will do so when time permits.

**Comment by brianmay** [comment]
I sometimes see "Connection lost - no network connection". Looking at the source this happens if ``isOnline()`` returns False.

https://github.com/dc297/mqttclpro/blob/01965085ad328d2f7983186e1f3a3adda2e01cb4/app/src/main/java/in/dc297/mqttclpro/MQTTService.java#L1089

So either the network never went down and this function sometimes returns a bad value, or the network did go down for a moment (which I doubt), resulting in the connection being lost, and it hasn't retried to connect.

**Comment by brianmay** [comment]
Seems to handle reconnecting to network fine. So I suspect the network never actually went down when I saw the "Connection lost - no network connection" message. As the network never went down, it is not going to come up, so it doesn't reconnect.

**Comment by brianmay** [comment]
The "Connection lost - no network connection" message is in the ``ConnectionLost`` function. Trying to work out what calls this? Maybe Android? https://github.com/dc297/mqttclpro/blob/01965085ad328d2f7983186e1f3a3adda2e01cb4/app/src/main/java/in/dc297/mqttclpro/MQTTService.java#L101

**Comment by brianmay** [comment]
Ok, so it seems that Android really is shutting the network down after the screen is off for about 4 to 5 minutes. This is despite having "Keep Wi-Fi on during sleep." set to always. So the disconnects aren't a problem in this project, the problem here is that the reconnects don't always work reliably.

**Comment by brianmay** [comment]
The phone also shuts down 3G too. Not just wifi.

**Comment by brianmay** [comment]
Looks like pinging an Android device isn't a good way of determining if the wifi is awake or not. This was confusing me into thinking that nothing I did was working.

I have found that if I set this application to be "Not optimized" in the battery optimization settings, it won't disconnect. Connection appears to be OK, even though the device stops responding to pings. https://support.google.com/pixelphone/answer/7015477?hl=en

There is probably another application that I should set to "Not optimized" to get pings to work.

**Comment by acefed** [comment]
Android disabling wifi might be a cause for network interruption, but the interrupted MQTT connection also happens to me when I stay connected via mobile data (OpenVPN), with no wifi activity at all. Today I saw "Unable to publish as we are not connected" while surfing the interwebs. In other words: With a working data connection. MQTT Client reconnects again when I open the app.

If you're digging into wifi connection details, logging the stats (writing them to a file) with Tasker for a period of time might help.

**Comment by brianmay** [comment]
Yes, there is still a problem here. Even if Android is disconnecting the network, it should still be possible to reconnect again after Android wakes up. As demonstrated by other programs that do reconnect (and hide the fact they were disconnected...). I suspect there is something wrong in the reconnect code.

**Comment by brianmay** [comment]
I notice that there are a few logging calls in the source, the best way forward would probably be to connect the phone to something like "adb logcat" to display log messages and try to reproduce the problem. tcpdump information could be useful too. There are also android apps to display logging but looks like these require Android be rooted - something I have not got around to doing. Reproducing the problem might be difficult, I have noticed that sometimes it reconnects perfectly.

**Comment by dc297** [comment]
Hey Brian

I have also started monitoring the issue closely. According to my research/knowledge, the issue is with Android's AlarmManager not notifying the app at the right time or not notifying the app at all.

Android generally skips sending the alarm to the app at the right time while the app is in background in order to save battery. Once an alarm is skipped, the app won't reconnect automatically. Maybe we need to find an alternative to Android's AlarmManager

**Comment by brianmay** [comment]
Ok, that makes sense... I was beginning to suspect something to do with the AlarmManager.

It should reconnect also when the connection comes up. Although it is possible that when it does try connecting there isn't a good network signal (I don't have any evidence here) causing it to fail, and it falls back to the faulty AlarmManager.

**Comment by acefed** [comment]
A faulty alarm manager sounds like a plausible cause for subscription/connection problems, but less so for the message "Unable to publish as we are not connected", when there is a working data connection active, handling browser traffic correctly at the same time. But I'm not an Android developer, so I could be completely wrong here.

I noticed other apps frequently using a visible notification icon, to keep processes alive (running in the foreground) on Android, as a workaround for the OS killing off background services when resources are needed. For example, see Tasker > Preferences > Monitor tab > Run in Foreground / Show Notofication Icon > Help

**Comment by dc297** [comment]
Hey I have pushed an update to the App. Now the app doesn't use AlarmManager any more, I have changed it to Handler. I have been testing it for the last 2 days and been working fine for me. 

Do check it out and let me know how it works for you.

Cheers!
Deepesh

**Comment by acefed** [comment]
Thanks! I have updated, testing in progress...

**Comment by acefed** [comment]
Unfortunately, when I have an active wifi connection but my screen is locked for some time, a MQTT subscription does not trigger an action in Tasker yet, when a message is sent. If I wake up the phone and try again, the action triggers correctly (without opening Tasker or MQTT Client).

So reconnection on a wake up works, but I never had a problem with that to begin with. In my case, reconnection on wake up always worked correctly.

Will test some more, to get some insight in the "Unable to publish as we are not connected" issue.

**Comment by dc297** [comment]
Hey have you disabled battery optimization for the app?

**Comment by acefed** [comment]
Yes, see above:

- Android power management is disabled for Tasker and Mqttclpro, including background services;
- Android set to Wifi: Always on, with no data limitation (one phone still disconnects wifi sometimes when locked/sleeping, other phone doesn't);

I tested on the phone which never disconnects wifi.

**Comment by dc297** [comment]
Could it be that Samsung has implemented some aggressive battery saving features which are denying Internet access to the app? 
Can you check logcat if the app is getting disconnected or not? Or if the app is receiving tasker actions or not.

**Comment by acefed** [comment]
That's certainly a possibility I can't rule out (yet). I use a local Mosquitto broker on my LAN, so when on wifi at home, no internet access is required. Other connections all go through OpenVPN, which could also be a source of errors. I haven't used logcat yet, but can look into that in a few days.

I'm also curious if @brianmay noticed an improvement in v2.3.0+. If so, chances are that my problem lies elsewhere, outside your app.

**Comment by brianmay** [comment]
I have noticed several times events fail to get through first thing in the morning.... But this also coincides with an update of the app during the night. It wouldn't surprise me if Android has a limitation where you can't automatically restart a service after an update. Perhaps not worth even mentioning here, however I decided to say so just in case other people have encountered the same issue.

Apart from that I have only had one disconnect issue since I turned off battery optimisation. So it might be a matter of me turning on battery optimisation again to test the reconnect works :-).

**Comment by brianmay** [comment]
Should also mention that I use OpenVPN too, also for when not at home. No problems so far. Note I did turn battery optimisation for OpenVPN too.

**Comment by acefed** [comment]
Good point, I did that too (some time ago, when I switched from OpenVPN Connect to OpenVPN for Android). Same goes for OpenVpn Tasker Plugin, just to be sure.

**Comment by acefed** [comment]
Haven't had time to look into this more thoroughly... still seeing "Unable to publish", but less so when I expect them. Sometimes as a surprise when I don't expect them. Very weird.

Without a real clue, my uneducated guess blames Android or Samsung software, fiddling with processes which don't use an active icon in the task bar.

**Comment by dc297** [comment]
Yeah, I thiink having an icon in the notification bar is the way forward. I think I'll add a persistent notification   .

**Comment by deepessh** [comment]
Hey guys..do we have any new observations on this after the new update?

**Comment by Bra1nK** [comment]
It's still happening due I believe to my Samsung S7's aggressive power saving. I think the permanent task bar notification is the way to go as demonstrated by other apps e.g. Tasker itself

**Comment by Bra1nK** [comment]
I monitor mine by using the Last Will feature to post the message "Lost" to a "Connection" topic

**Comment by brianmay** [comment]
I notice that Android 8.0.0 lists tasks running in the background. On my phone Tasker, CSipSimple and LastPass are listed. But mqttclpro is omitted from this list. I suspect mqttclpro needs to behave like the other background tasks, but I have no idea how this is done. Maybe this is the permanent task bar notification, not sure.

**Comment by deepessh** [comment]
I didn't want to introduce an annoying notification. But I think we don't have a choice. Let me see what's feasible.

**Comment by Bra1nK** [comment]
Make it optional, not everyone one needs a persistent connection, I only need it when at home.

**Comment by deepessh** [comment]
Yeah, that's what I was thinking.

**Comment by deepessh** [comment]
Hey guys. I have added notification option in v4.3.0. Please test if it helps.

**Comment by Bra1nK** [comment]
Now on v 4.4.1 and been testing the persistent notification which although it improves things in some respects (only had 2 disconnects overnight rather than the usual 5 or 6 whilst I wasn't using the phone myself), doesn't seem any better whilst I'm using the phone for other stuff. 
It may not be related but just now I was outside so WiFi disconnected but when I came back in and reconnected the app failed to reconnect to the broker. 
![screenshot_20171123-151309](https://user-images.githubusercontent.com/3098712/33180252-fbf48846-d063-11e7-9055-e216686a641a.png)
with a failed to connect message saying it would try again at a time that was already past, in fact it was another 5 minutes before it actually reconnected then all the stored messages were sent.

**Comment by brianmay** [comment]
I have found that sometimes the persistent icon isn't displayed and messages get missed. Not 100% sure yet, might be related to what happens after an upgrade. After opening the app manually, things start working correctly with no sign of any problem.

**Comment by deepessh** [comment]
Hey Brian

That failed to connect message is not reliable.
The paho client library doesn't support callbacks in case of reconnect failure. So it is difficult to show the correct times for retries. It is, however, on my list.

Currently, the way reconnect works is it starts with retrying every second and doubles retry time whenever it fails to reconnect up until we reach 128 seconds. Then it will keep on trying every 128 seconds.

I think we definitely see an improvement after introducing persistent notification.

Ideally, that notification shouldn't be removed for any reason.

Yes, let's monitor more.

Cheers!
Deepesh

**Comment by bertbesser** [comment]
I encounter disconnects with version 4.4.1, i.e. with persistent notification. I'm on Android 7.1.2, Lineage OS 14.1 to be precise.

**Comment by brianmay** [comment]
I suspect I have had some problems recently, but is relatively rare. As a result, last time it caught me by surprise and I didn't try to check if the service was running.

**Comment by deepessh** [comment]
I have been monitoring this myself and found that more than 50 disconnects occur throughout the day. The huge number might be due to the incompatibility between my phone's and router's wifi standards. 

However, every time I open the app I can see that we are connected to the broker. I am not sure what can be done to mitigate the disconnects in the first place. Would love to hear some suggestions.

**Comment by haydster7** [comment]
Is that around 50 disconnects per day even with battery optimisations off and persistent notification on?

I regularly get disconnects, but the main issue is no matter what it won't reconnect on its own. The only way I can reconnect is by changing a setting in the broker settings, then changing it back forcing a reconnect.

**Comment by deepessh** [comment]
@haydster7 this is a different issue. Not all the users are facing this. Can you open a new issue along with the details that might help in debugging?

**Comment by bertbesser** [comment]
After turning off battery optimization for the app and configuring the wireless connection to be always on, I suddenly encounter only very few dis-/reconnects, if any at all. (My fault for not tuning the android config earlier.)

**Comment by brianmay** [comment]
Had a issue this morning where it said it was connected, but not receiving any messages. I had to kill and restart it for it to wake up. First time this has happened in a while.

**Comment by haydster7** [comment]
It's not just me then
I was waiting until I was certain to report it.
Pressing the save button manually as in #52 did not resolve the issue in this instance.
Do you have SSL enabled?

**Comment by stipe76** [comment]
I need to go in and click save for it to reconnect

On Jan 3, 2018 08:55, "Hayden Moulds" <notifications@github.com> wrote:

It's not just me then
I was waiting until I was certain to report it.
Pressing the save button manually did not resolve the issue in this
instance.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/dc297/mqttclpro/issues/19#issuecomment-354885430>, or mute
the thread

**Comment by haydster7** [comment]
Stipe, do you have SSL enabled?

**Comment by stipe76** [comment]
No, not as yet

**Comment by deepessh** [comment]
Hey guys. How frequent is the issue?

**Comment by Bra1nK** [comment]
In my case 6-12 times overnight and up to 100 times during the day

**Comment by deel77** [comment]
I am using Automate (Tasker alternative) to publish MQTT message when my phone starts ringing, but I am also having problem with mqttclpro that after some inactivity with my phone (OnePlus 3) first message is not received by MQTT broker. I am not subscribed to any topic, therefore I think it should just connect to broker on message send, but somehow it does not happen. But it starts working right after first message.

I have both apps Mqttclpro and Automate set to Not optimize for battery optimization. And I am using SSL connection to cloudmqtt.com (mosquitto broker).

Does Mqttclpro store messages to be sent in case of some connection problem and resend them after successful connection? Or can it be some problem with between app comunication? As Automate app has log, and there is call for mqtt plugin.

**Comment by jantenhove** [comment]
I noticed the same problem of not publishing the first time. I use mqttclpro only for publishing. Is it possible to disconnect after each publish (and connect on each publish, with a retry interval)?

**Comment by stipe76** [comment]
I am getting regular disconnects many times a day, and it won't reconnect until I go into broker settings and click Save. Pretty annoying having to do this every so often. Is there any way to fix this?

**Comment by deepessh** [comment]
@stipe76 
Only a few users are facing this issue. Have you observed anything specific which might help in debugging the issue?

Let's open a new issue and monitor this issue closely. Please share the logcat as well.

Regards

**Comment by NewUser9** [comment]
Disconnects only happen on Samsung devices. I have two samsung phones both of them disconnects when keep-alive time limit is reached. 
For example: If I set keep alive to 60 seconds it will reconnect every 60 seconds.

EDIT:
Today I received this message from Samsung smart manager app.

EDIT2:
App v4.4.4 There seems to be an problem with alarmmanager. 


![screenshot_20180528-070058__01](https://user-images.githubusercontent.com/37405841/40594129-d829c2f4-624a-11e8-89dd-c4b4ada1b706.jpg)

**Comment by deepessh** [comment]
@NewUser9 Can you share logcat for the time when disconnects happen?

The app seems to be crashing due to some reason.

**Comment by NewUser9** [comment]
@dc297 as of now I'm not facing disconnects for some reasons I'll send you the log if it happens again.

**Comment by jsetton** [comment]
Related to Samsung disconnects, I found out recently, after upgrading my Galaxy S7 to Android 8.0, that there are actually two configuration locations where an app needs to be added in order to exclude it completely from any battery optimization mechanism. Since I made the change, it seems the disconnects aren't happening anymore.

Here are the two configuration locations:

- System Settings > Apps > Special access (Dots menu) > Optimize battery usage > All Apps (Drop-down menu) > [Select MQTT Client app]

- System Settings > Device Maintenance > Battery (Icon) > Unmonitored apps (Bottom) > Add apps (Bottom) > [Turn off MQTT Client app]

**Comment by deepessh** [comment]
@jsetton thanks for sharing. This should help other users.

**Comment by Bra1nK** [comment]
Just got round to checking those locations and the MQTT client was selected in the 1st one (i.e optimised) so I deselected it and it now appears in the unmonitored apps list. Will now monitor for disconnects

**Comment by fuslwusl** [comment]
I have the same problem. When display locks the connection gets lost and no messages will be received.
I found the following info that may be helpful for developers:
https://stackoverflow.com/questions/50132566/wake-up-device-once-it-enters-in-to-doze-mode

**Comment by Bra1nK** [comment]
Things have improved since adding MQTT client to the unmonitored apps but I still get disconnects just not as many as before

**Comment by fuslwusl** [comment]
Here are some infos I found that may be helpful to solve the problem:
https://github.com/eclipse/paho.mqtt.android/issues/195
https://stackoverflow.com/questions/50132566/wake-up-device-once-it-enters-in-to-doze-mode

**Comment by fuslwusl** [comment]
The problem for all disconnects is the Android doze mode - https://developer.android.com/training/monitoring-device-state/doze-standby
Powering the device or display always on are ways to avoid disconnects and receive mqtt messages without delay. Another way could be a server solution using Google cloud messages (firebase) to send the mqtt message and topic info to the Android app instead of a direct mqtt transfer.
Does anybody know better ways to receive reliable real-time mqtt messages?

**Comment by Bra1nK** [comment]
I recently upgraded my phone from Samsung S7 Edge to Samsung S9 plus and the disconnects have ceased and the only connection lost (Last Will) messages I'm seeing are the ones I'd expect (i.e. when I leave my WiFi).  Very strange seeing as they're both on Android 8.0 but welcome nevertheless.
